File: apps/wear_os/app/src/main/java/com/fallguardian/CompanionEnrollmentValidation.kt
package com.fallguardian

import java.time.Instant
import java.time.format.DateTimeParseException

/**
 * Pure validation for an inbound `/companion_enrollment` Data Layer message,
 * kept free of Android/Wearable/org.json APIs so it runs in plain JUnit tests
 * (see CompanionEnrollmentValidationTest.kt) the same way FallAlgorithm is
 * tested — `org.json.JSONObject` throws "not mocked" under the JVM unit-test
 * classpath, so this takes a plain Map instead; CompanionEnrollmentClient
 * converts the parsed JSONObject to a Map before calling in.
 *
 * Per docs/COMPANION_ENROLLMENT.md §3 ("Watch" responsibilities): "verify
 * platform and expiration before the network call" — this is that check, run
 * before CompanionEnrollmentClient ever touches the network.
 */
data class ValidatedEnrollment(val enrollmentToken: String, val expiresAt: Instant)

object CompanionEnrollmentValidation {
    fun validate(message: Map<String, Any?>, expectedPlatform: String, now: Instant): ValidatedEnrollment? {
        if (message["type"] != "companionEnrollment") return null
        if ((message["schemaVersion"] as? Number)?.toInt() != 1) return null
        if (message["platform"] != expectedPlatform) return null

        val token = message["enrollmentToken"] as? String
        if (token.isNullOrBlank()) return null

        val expiresAtRaw = message["expiresAt"] as? String
        if (expiresAtRaw.isNullOrBlank()) return null

        val expiresAt = try {
            Instant.parse(expiresAtRaw)
        } catch (_: DateTimeParseException) {
            return null
        }

        if (!expiresAt.isAfter(now)) return null

        return ValidatedEnrollment(token, expiresAt)
    }
}

File: apps/watchos/FallGuardian/FallGuardian Watch App/CompanionEnrollmentValidation.swift
// CompanionEnrollmentValidation.swift
// Fall Guardian — watchOS
//
// Pure validation for an inbound companionEnrollment WatchConnectivity
// message, kept free of WatchConnectivity/Security/Foundation-networking
// imports so it compiles and runs standalone via `swiftc` for deterministic
// tests (see FallGuardianTests/CompanionEnrollmentValidationExecutableTests.swift
// and `make test`), the same pattern used by FallAlgorithm.swift.
//
// Per docs/COMPANION_ENROLLMENT.md §3 ("Watch" responsibilities): "verify
// platform and expiration before the network call" — this is that check,
// run before CompanionEnrollmentClient ever touches the network.

import Foundation

enum CompanionEnrollmentValidation {
    struct ValidatedEnrollment {
        let enrollmentToken: String
        let expiresAt: Date
    }

    /// Returns the validated token and expiry, or `nil` if the message is
    /// malformed, addressed to a different platform, or already expired.
    static func validate(
        _ message: [String: Any],
        expectedPlatform: String,
        now: Date = Date()
    ) -> ValidatedEnrollment? {
        guard message["type"] as? String == "companionEnrollment",
              message["schemaVersion"] as? Int == 1,
              message["platform"] as? String == expectedPlatform,
              let token = message["enrollmentToken"] as? String,
              !token.isEmpty,
              let expiresAtString = message["expiresAt"] as? String,
              let expiresAt = ISO8601DateFormatter().date(from: expiresAtString) else {
            return nil
        }

        guard expiresAt > now else { return nil }

        return ValidatedEnrollment(enrollmentToken: token, expiresAt: expiresAt)
    }
}
